torch/distributed/_sharding_spec/api.py

```diff
@@ -96,3 +96,173 @@ def placement(self) -> ShardPlacement:
+            rank = len(shard.shard_offsets)
+
+        self._validate_non_overlapping(shards)
+
+        self._shards = shards
+
+    @staticmethod
+    def _validate_non_overlapping(shards: List[Shard]):
+        """
+        Ensures none of the shards overlap with each other.
+        """
+        for i in range(len(shards)):
+            for j in range(i + 1, len(shards)):
+                if GenericShardingSpec._check_shard_pair_overlap(shards[i], shards[j]):
+                    raise ValueError(f'Shards {shards[i]} and {shards[j]} overlap')
+
+    @staticmethod
+    def _check_shard_pair_overlap(shard1: Shard, shard2: Shard):
+        """
+        Checks if two shards overlap.
+        """
+
+        # For each dim of each shard, check if one shard resides on the other
+        # end of second shard with respect to that dim. As an example for a 2D
+        # shard, we would check if one shard is above or on the left of the
+        # other shard.
+        ndims = len(shard1.shard_offsets)
+        for i in range(ndims):
+            if shard1.shard_offsets[i] >= shard2.shard_offsets[i] + shard2.shard_lengths[i]:
+                return False
+            if shard2.shard_offsets[i] >= shard1.shard_offsets[i] + shard1.shard_lengths[i]:
+                return False
+
+        return True
+
+    @property
+    def shards(self):
+        return self._shards
+
+    def check_tensor(self, tensor: torch.Tensor) -> None:
+        """
+        Checks if the sharding spec is compatible with the provided tensor.
+
+        Args:
+            tensor(torch.Tensor): Tensor to verify.
+        Raises:
+            ``ValueError`` if not compatible.
+        """
+
+        # If the tensor's volume matches the total volume of all shards and
+        # all shard boundaries are within tensor dims, we have a compatible
+        # sharding spec for this tensor. Note that we have already verified
+        # we don't have overlapping shards.
+        tensor_rank = len(tensor.size())
+        shards_rank = len(self._shards[0].shard_offsets)
+        if tensor_rank != shards_rank:
+            raise ValueError(f'Rank of tensor is {tensor_rank}, but shards rank is {shards_rank}')
+
+        total_shard_volume = 0
+        tensor_dims = tensor.size()
+        for shard in self._shards:
+            shard_volume = 1
+            for i, shard_length in enumerate(shard.shard_lengths):
+                shard_volume *= shard_length
+                if shard.shard_offsets[i] + shard.shard_lengths[i] > tensor_dims[i]:
+                    raise ValueError(
+                        f'Shard offset {shard.shard_offsets[i]} and length'
+                        f'{shard.shard_lengths[i]} exceeds tensor dim: {tensor_dims[i]} for shard {shard}')
+            total_shard_volume += shard_volume
+
+        tensor_volume = 1
+        for size in tensor_dims:
+            tensor_volume *= size
+
+        if total_shard_volume != tensor_volume:
```

Comment:
If `total_shard_volum` < `tensor_volume`, better to help identify the gaps that are not covered by the shards. Otherwise it can be very hard for the developers to eyeball all the shards. Therefore, it can be valuable to create a separate util function for providing more informative error message.